File: src/commands/file-commands.ts
import { invoke } from '@tauri-apps/api/core';

// 文件操作相关的类型定义
export interface FileResult<T = any> {
  success: boolean;
  data?: T;
  error?: string;
}

export interface FileInfo {
  path: string;
  size: number;
  modified: string;
  isFile: boolean;
  isDirectory: boolean;
}

export interface DirectoryEntry {
  name: string;
  isFile: boolean;
  isDirectory: boolean;
}

// 文件存在性检查
export async function fileExists(path: string): Promise<FileResult<boolean>> {
  try {
    const result = await invoke<boolean>('file_exists', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileDirExists(path: string): Promise<FileResult<boolean>> {
  try {
    const result = await invoke<boolean>('file_dir_exists', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

// 文件读取操作
export async function fileRead(path: string): Promise<FileResult<string>> {
  try {
    const result = await invoke<string>('file_read', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileReadBytes(path: string): Promise<FileResult<Uint8Array>> {
  try {
    const result = await invoke<Uint8Array>('file_read_bytes', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

// 文件写入操作
export async function fileWrite(path: string, content: string): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_write', { path, content });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileWriteBytes(path: string, content: Uint8Array): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_write_bytes', { path, content });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileAppend(path: string, content: string): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_append', { path, content });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

// 目录操作
export async function fileCreateDir(path: string): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_create_dir', { path });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileDelete(path: string): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_delete', { path });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileCopy(from: string, to: string): Promise<FileResult<number>> {
  try {
    const result = await invoke<number>('file_copy', { from, to });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileMove(from: string, to: string): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_move', { from, to });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

// 文件信息操作
export async function fileSize(path: string): Promise<FileResult<number>> {
  try {
    const result = await invoke<number>('file_size', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileModified(path: string): Promise<FileResult<string>> {
  try {
    const result = await invoke<string>('file_modified', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileListDir(path: string): Promise<FileResult<string[]>> {
  try {
    const result = await invoke<string[]>('file_list_dir', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

// JSON 操作
export async function fileReadJson<T = any>(path: string): Promise<FileResult<T>> {
  try {
    const result = await invoke<T>('file_read_json', { path });
    return { success: true, data: result };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function fileWriteJson<T = any>(path: string, data: T): Promise<FileResult<void>> {
  try {
    await invoke<void>('file_write_json', { path, data });
    return { success: true };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : String(error) };
  }
}

// 实用工具函数
export async function ensureDir(path: string): Promise<FileResult<void>> {
  const dirExistsResult = await fileDirExists(path);
  if (!dirExistsResult.success) {
    return { success: false, error: dirExistsResult.error };
  }
  
  if (!dirExistsResult.data) {
    return await fileCreateDir(path);
  }
  
  return { success: true };
}

export async function readJsonFile<T = any>(path: string, defaultValue: T): Promise<FileResult<T>> {
  const result = await fileReadJson<T>(path);
  if (!result.success) {
    // 如果文件不存在或读取失败，返回默认值
    if (result.error?.includes('No such file') || result.error?.includes('读取 JSON 文件失败')) {
      return { success: true, data: defaultValue };
    }
    return result;
  }
  return result;
}

export async function writeJsonFile<T = any>(path: string, data: T): Promise<FileResult<void>> {
  const dirPath = path.substring(0, path.lastIndexOf('/'));
  if (dirPath) {
    const dirResult = await ensureDir(dirPath);
    if (!dirResult.success) {
      return dirResult;
    }
  }
  
  return await fileWriteJson(path, data);
}

File: src-tauri/src/commands/file_commands.rs
use std::io::Write;

// 文件存在性检查命令
#[tauri::command]
pub fn file_exists(path: String) -> Result<bool, String> {
  match std::path::Path::new(&path).exists() {
    true => Ok(true),
    false => Ok(false),
  }
}

#[tauri::command]
pub fn file_dir_exists(path: String) -> Result<bool, String> {
  match std::path::Path::new(&path).exists() {
    true => Ok(std::path::Path::new(&path).is_dir()),
    false => Ok(false),
  }
}

// 文件读取命令
#[tauri::command]
pub fn file_read(path: String) -> Result<String, String> {
  std::fs::read_to_string(&path)
    .map_err(|e| format!("读取文件失败: {}", e))
}

#[tauri::command]
pub fn file_read_bytes(path: String) -> Result<Vec<u8>, String> {
  std::fs::read(&path)
    .map_err(|e| format!("读取文件失败: {}", e))
}

// 文件写入命令
#[tauri::command]
pub fn file_write(path: String, content: String) -> Result<(), String> {
  std::fs::write(&path, content)
    .map_err(|e| format!("写入文件失败: {}", e))
}

#[tauri::command]
pub fn file_write_bytes(path: String, content: Vec<u8>) -> Result<(), String> {
  std::fs::write(&path, content)
    .map_err(|e| format!("写入文件失败: {}", e))
}

#[tauri::command]
pub fn file_append(path: String, content: String) -> Result<(), String> {
  let mut file = std::fs::OpenOptions::new()
    .create(true)
    .append(true)
    .open(&path)
    .map_err(|e| format!("打开文件失败: {}", e))?;
  file.write_all(content.as_bytes())
    .map_err(|e| format!("追加文件失败: {}", e))
}

// 目录操作命令
#[tauri::command]
pub fn file_create_dir(path: String) -> Result<(), String> {
  if !std::path::Path::new(&path).exists() {
    std::fs::create_dir_all(&path)
      .map_err(|e| format!("创建目录失败: {}", e))
  } else {
    Ok(())
  }
}

#[tauri::command]
pub fn file_delete(path: String) -> Result<(), String> {
  if std::path::Path::new(&path).exists() {
    std::fs::remove_file(&path)
      .map_err(|e| format!("删除文件失败: {}", e))
  } else {
    Ok(())
  }
}

#[tauri::command]
pub fn file_copy(from: String, to: String) -> Result<u64, String> {
  std::fs::copy(&from, &to)
    .map_err(|e| format!("复制文件失败: {}", e))
}

#[tauri::command]
pub fn file_move(from: String, to: String) -> Result<(), String> {
  std::fs::rename(&from, &to)
    .map_err(|e| format!("移动文件失败: {}", e))
}

// 文件信息命令
#[tauri::command]
pub fn file_size(path: String) -> Result<u64, String> {
  let metadata = std::fs::metadata(&path)
    .map_err(|e| format!("获取文件元数据失败: {}", e))?;
  Ok(metadata.len())
}

#[tauri::command]
pub fn file_modified(path: String) -> Result<String, String> {
  let metadata = std::fs::metadata(&path)
    .map_err(|e| format!("获取文件元数据失败: {}", e))?;
  let modified = metadata.modified()
    .map_err(|e| format!("获取修改时间失败: {}", e))?;
  Ok(format!("{:?}", modified))
}

#[tauri::command]
pub fn file_list_dir(path: String) -> Result<Vec<String>, String> {
  let entries = std::fs::read_dir(&path)
    .map_err(|e| format!("读取目录失败: {}", e))?;
  let mut names = Vec::new();
  
  for entry in entries {
    let entry = entry.map_err(|e| format!("读取目录项失败: {}", e))?;
    if let Some(name) = entry.file_name().to_str() {
      names.push(name.to_string());
    }
  }
  
  Ok(names)
}

// JSON 操作命令
#[tauri::command]
pub fn file_read_json(path: String) -> Result<serde_json::Value, String> {
  let content = std::fs::read_to_string(&path)
    .map_err(|e| format!("读取 JSON 文件失败: {}", e))?;
  let data: serde_json::Value = serde_json::from_str(&content)
    .map_err(|e| format!("解析 JSON 失败: {}", e))?;
  Ok(data)
}

#[tauri::command]
pub fn file_write_json(path: String, data: serde_json::Value) -> Result<(), String> {
  let content = serde_json::to_string_pretty(&data)
    .map_err(|e| format!("序列化 JSON 失败: {}", e))?;
  std::fs::write(&path, content)
    .map_err(|e| format!("写入 JSON 文件失败: {}", e))?;
  Ok(())
}